extension/test/e2e/credit-card.spec.js

```diff
@@ -0,0 +1,118 @@
+    { name: 'Elo', creditCardNumber: '5066 9911 1111 1118' },
+    { name: 'Hipercard', creditCardNumber: '6062 8288 8866 6688' },
+    { name: 'JCB', creditCardNumber: '3569 9900 1009 5841' },
+    { name: 'Mastercard', creditCardNumber: '5101 1800 0000 0007' },
+    { name: 'UATP', creditCardNumber: '1354 1001 4004 955', creditCardDate: '06/22' },
+    { name: 'VISA', creditCardNumber: '4166 6766 6766 6746' },
+    { name: 'Visa Electron', creditCardNumber: '4001 0200 0000 0009' },
+    { name: 'V Pay', creditCardNumber: '4013 2500 0000 0000 006' },
+  ]
+
+  // note: ngrok should be restarted for every test case, otherwise there will be
+  // 429 Too Many Requests error. This is due to the limit of maximum opened HTTP connections,
+  // which is 40 connections at the same time.
+  beforeEach(async () => {
+    const fileServer = new nodeStatic.Server()
+    routes['/make-payment-form'] = async (request, response) => {
+      fileServer.serveFile('./test/e2e/fixtures/make-payment-form.html', 200, {}, request, response)
+    }
+    ctpClient = ctpClientBuilder.get()
+    await iTSetUp.initServerAndExtension({ ctpClient, routes, testServerPort: 8080 })
+    browser = await puppeteer.launch({
+      headless: true,
+      ignoreHTTPSErrors: true,
+      args: ['--disable-web-security', '--disable-features=IsolateOrigins,site-per-process']
+    })
+  })
+
+  afterEach(async () => {
+    await iTSetUp.stopRunningServers()
+    await browser.close()
+  })
+
+  creditCards.forEach(({
+                         name, creditCardNumber, creditCardDate = '03/30', creditCardCvc = '737'
+                       }) => {
+    // eslint-disable-next-line no-template-curly-in-string
+    it(`when issuer is ${name} and number is ${creditCardNumber}, ` +
+      'it should successfully finish the payment with credit card',
+      async () => {
+        const page = await browser.newPage()
+        const baseUrl = process.env.API_EXTENSION_BASE_URL
+
+        const getOriginKeysRequestDraft = {
+          originDomains: [
+            baseUrl
+          ]
+        }
+        const paymentDraft = {
+          amountPlanned: {
+            currencyCode: 'EUR',
+            centAmount: 1000,
+          },
+          paymentMethodInfo: {
+            paymentInterface: c.CTP_ADYEN_INTEGRATION
+          },
+          custom: {
+            type: {
+              typeId: 'type',
+              key: c.CTP_PAYMENT_CUSTOM_TYPE_KEY
+            },
+            fields: {
+              getOriginKeysRequest: JSON.stringify(getOriginKeysRequestDraft)
+            }
+          }
+        }
+
+        const { body: payment } = await ctpClient.create(ctpClient.builder.payments, paymentDraft)
+        const { getOriginKeysResponse: getOriginKeysResponseString } = payment.custom.fields
+        const getOriginKeysResponse = await JSON.parse(getOriginKeysResponseString)
+
+        // Make payment:
+        const makePaymentFormPage = new MakePaymentFormPage(page, baseUrl)
```

Comment:
Why this is not called `CardComponent` as Adyen calls it that way? https://docs.adyen.com/payment-methods/cards/web-component

Think those e2e tests as documentation. It is always good to use the right naming and easy to reason code.

do we need to pass a page, browser, static file router ext in here, maybe it should be part of the component, otherwise, there is no value doing a decoupling as you did. So CardComponent will manage everything related to client-side and fixture will only do the extension call part (like the customer will do the same )?  

new CardComponent() ? 

Here I guess we need to define a strategy, another option for me we could also remove decoupling in here and write the puppeteer operations on the same test instead of moving them into the page objects folder which is confusing. If we are doing decoupling I would ask for a better implementation, so this means you need to structure the tests with this goal in mind.

Reply:
so reusability for puppeteer operations is not a problem in here, it should be clean. check the design below, the code is passing page as a constructor but not using anywhere else, which is actually a state of the MakePaymentFormPage which is managed by there, maybe same for baseUrl. 

https://github.com/commercetools/commercetools-adyen-integration/blob/ca36b206c94043b9acc47ec8691809b28e127a40/extension/test/e2e/credit-card.spec.js#L70-L73

So check this comment again: 
> do we need to pass a page, browser, static file router ext in here, maybe it should be part of the component, otherwise, there is no value doing a decoupling as you did.

Here another case, so this should be a method. 
https://github.com/commercetools/commercetools-adyen-integration/blob/ca36b206c94043b9acc47ec8691809b28e127a40/extension/test/e2e/credit-card-3ds-native.spec.js#L133-L136

So my comment is neither implemented nor understood.

> If we are doing decoupling I would ask for a better implementation, so this means you need to structure the tests with this goal in mind.

Also, please better names... you are not using the domain language in the naming, in the end tests are not only for something to evaluate the code, IMO especially e2e tests should also be meaningful. 

Let me know how we could improve the existing flow especially readability in tests. (should not be hard to do for you)

Reply:
https://github.com/commercetools/commercetools-adyen-integration/pull/323 Thanks :) Now I understand the comment. I should split the big test method into smaller methods.

Reply:
https://github.com/commercetools/commercetools-adyen-integration/pull/321/commits/2a23d0c0c3665f135a628a3311dd04129ca838e6